### PR TITLE
Build and test Docker on IBM Power and Z using gccgo. Enable CI on Po…

### PR DESCRIPTION
**Pull request by brahmaroutu**
…wer and Z.

Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

This PR will allow us to setup a Docker CI runs Power and Z platforms.

### CODE REVIEW

**Comment by brahmaroutu** [comment]
ping @tianon @jfrazelle Please review the PR, this will hopefully help us setup CI for Power and Z.

**Comment by hqhq** [comment]
@brahmaroutu I have some questions,
- If you use images like `ppc64le/busybox` for the official busybox image on ppc64le, who and how should people maintain images for ppc64le?
- Currently is that depend on the user "ppc64le"?
- Would there be official and personal images under `ppc64le`?

**Comment by brahmaroutu** [comment]
ping @tianon @jfrazelle 
@hqhq ppc64le should only have hello-world and busybox and we maintain that.

**Comment by LK4D4** [comment]
ping @tianon 
Pls take a look.

File: Dockerfile.ppc64le [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# This file describes the standard way to build Docker, using docker
+#
+# Usage:
+#
+# # Assemble the full dev environment. This is slow the first time.
+# docker build -t docker -f Dockerfile.gccgo .
```
Comment (tianon) [inline Dockerfile.ppc64le]: This should be `.ppc64le` and should mention that this will _only_ work on a `ppc64le` machine.

File: Dockerfile.ppc64le [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#
+
+FROM ppc64le/gcc:5.2
+ENV CGO_ENABLED 1
+ENV USE_GCCGO  1
```
Comment (tianon) [inline Dockerfile.ppc64le]: What are these `ENV` vars here for?

File: Dockerfile.ppc64le [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+# Install lxc
+ENV LXC_VERSION 1.1.2
+RUN mkdir -p /usr/src/lxc \
+	&& curl -sSL https://linuxcontainers.org/downloads/lxc/lxc-${LXC_VERSION}.tar.gz | tar -v -C /usr/src/lxc/ -xz --strip-components=1
+RUN cd /usr/src/lxc \
+	&& ./configure \
+	&& make \
+	&& make install \
+	&& ldconfig
```
Comment (tianon) [inline Dockerfile.ppc64le]: This is a perfect example of why I'm really, really, _really_ uncomfortable with having additional copies of the main `Dockerfile`.  This is already outdated, and even if it'd merged before the previous one that removed this, it likely wouldn't have been updated, because most contributors don't even _think_ to check other files when need to update the main `Dockerfile` (or any other file like this).

File: hack/make/.ensure-images [hunk 4]
```diff
@@ -0,0 +1,9 @@
+for image in `docker images | awk '{print $1}'`; do
+if ( [ -z "${image##ppc64le/busybox}" ] || [ -z "${image##s390x/busybox}" ] ); then
+	docker tag $image busybox:latest
+	docker rmi $image
+elif ( [ -z "${image##ppc64le/hello-world}" ] || [ -z "${image##s390x/hello-world}" ] ); then
+	docker tag $image:frozen hello-world:frozen
+	docker rmi $image:frozen
+fi
+done
```
Comment (tianon) [inline hack/make/.ensure-images]: This script is really hacky, and makes adding more architectures or more test images even more of a pain than it already is going to be -- can you think of some way we could centralize more of this in some way?  Ideally, we should only have to specify the mapping of "this organization account on the Hub corresponds to the top-level official images for this specific architecture" in a single place and have things like this fixed for us more naturally.

I also don't understand why this PR still has new `Dockerfile`s in it when I've said many times before now that I'd like to see CI working on these architectures _without_ using Docker to build (ie, `AUTO_GOPATH=1 ./hack/make.sh ...` the same way new systems have to be bootstrapped and the way packagers compile Docker) before we try building and testing inside Docker.  Since we still don't have multi-arch support in the Hub, I don't think it's a great idea for us to rush into any of this and get too stuck on one particular implementation.  The smaller we can make the changes for multi-arch, the better I'll feel about the project being able to support it in the long term.  This PR as-is is going to add pretty significant maintenance burden on the project with each new architecture we try to add.

File: hack/make/gccgo [hunk 5]
```diff
@@ -14,10 +14,12 @@ go build -compiler=gccgo \
 	-o "$DEST/$BINARY_FULLNAME" \
 	"${BUILDFLAGS[@]}" \
 	-gccgoflags "
+		-O0
```
Comment (tianon) [inline hack/make/gccgo]: What's the justification for this?

File: hack/make/gccgo [hunk 6]
```diff
@@ -14,10 +14,12 @@ go build -compiler=gccgo \
...
 		-g
 		$EXTLDFLAGS_STATIC
 		-Wl,--no-export-dynamic
 		-ldl
+		-pthread
```
Comment (tianon) [inline hack/make/gccgo]: Again, why is this necessary?  A little explanation in a comment here in the script would go a long way, if there's a good reason for this to be included.  If there's not a good reason for this to be included, then I think we shouldn't. :wink:

File: integration-cli/docker_cli_daemon_test.go [hunk 7]
```diff
@@ -1665,6 +1665,8 @@ func (s *DockerDaemonSuite) TestDaemonRestartWithContainerWithRestartPolicyAlway
 }
 
 func (s *DockerDaemonSuite) TestDaemonWideLogConfig(c *check.C) {
+	testRequires(c, x86_64)
```
Comment (tianon) [inline integration-cli/docker_cli_daemon_test.go]: Why does testing arguments to the `json-file` log driver require `x86_64`?

File: integration-cli/requirements.go [hunk 8]
```diff
@@ -170,6 +170,16 @@ var (
 		},
 		"Test cannot be run when remapping root",
 	}
+	x86_64 = testRequirement{
+		func() bool {
+			out, err := exec.Command("uname", "-p").Output()
+			if err != nil && (strings.Contains(string(out), "x86_64") || strings.Contains(string(out), "i386")) {
```
Comment (tianon) [inline integration-cli/requirements.go]: I think this test is poorly named, and that there's no reason why we should have to exclude tests in this way either at this point -- so far, we aren't using any images that can't really be replicated on other Linux platforms without too much trouble.

**Comment by tianon** [comment]
@hqhq the `ppc64le` and `s390x` namespaces are currently in IBM's+my control, and the only images pushed there should essentially only be rebuilds of the top-level images :+1:

File: Dockerfile.ppc64le (hunk 3 above)
Comment (duglin) [inline Dockerfile.ppc64le]: How would you like to solve this issue?

File: Dockerfile.ppc64le (hunk 1 above)
Comment (brahmaroutu) [inline Dockerfile.ppc64le]: fixed.

File: Dockerfile.ppc64le (hunk 2 above)
Comment (brahmaroutu) [inline Dockerfile.ppc64le]: fixed.

File: Dockerfile.ppc64le (hunk 3 above)
Comment (brahmaroutu) [inline Dockerfile.ppc64le]: Fixed.

File: hack/make/.ensure-images (hunk 4 above)
Comment (brahmaroutu) [inline hack/make/.ensure-images]: We probably not need this once multi-arch support is in place. Hope this solution suffice until then. I prefer not to code generically for more than busybox, hello-world if you agree.

File: hack/make/gccgo (hunk 5 above)
Comment (brahmaroutu) [inline hack/make/gccgo]: Fixed, taking it out. Only reason is to speed up the compilation.

File: hack/make/gccgo (hunk 6 above)
Comment (brahmaroutu) [inline hack/make/gccgo]: fixed.

File: integration-cli/docker_cli_daemon_test.go (hunk 7 above)
Comment (brahmaroutu) [inline integration-cli/docker_cli_daemon_test.go]: Fixed, the test is passing now, at some point in time this was failing on Power/Z consistently. I agree with you that this test has nothing to do with Architecture.

File: integration-cli/requirements.go (hunk 8 above)
Comment (brahmaroutu) [inline integration-cli/requirements.go]: No need for this requirement anymore.

**Comment by brahmaroutu** [comment]
@tianon Please review, code changed as per your comments.

**Comment by crosbymichael** [comment]
LGTM

**Comment by duglin** [comment]
ping @tianon for final review

**Comment by tianon** [comment]
Yeah, this will do for now, especially since #17803 merged already.  I'll work with @jfrazelle post-merge to figure out a better way for us to manage/maintain multiple platforms like this cleaner.

LGTM

**Comment by brahmaroutu** [comment]
Awesome. Thanks @crosbymichael @tianon

File: integration-cli/docker_cli_run_unix_test.go [hunk 9]
```diff
@@ -238,7 +238,7 @@ func (s *DockerSuite) TestRunOOMExitCode(c *check.C) {
 	select {
 	case err := <-errChan:
 		c.Assert(err, check.IsNil)
-	case <-time.After(30 * time.Second):
+	case <-time.After(600 * time.Second):
```
Comment (hqhq) [inline integration-cli/docker_cli_run_unix_test.go]: @brahmaroutu Any idea why this case might takes so long? I found on our CI server, this case takes almost 400s, though it's passed, and it's a x86-64 server.

Comment (brahmaroutu) [inline integration-cli/docker_cli_run_unix_test.go]: @hqhq as far as I remember gccgo(glibc) bootstrap and growing size of the binary was causing the extra time.
